### ISSUE TITLE
vmap to build gaussian profile images (mainly drawing branch)

### ISSUE THREAD

**Issue by jecampagne** [GalSim-developers/JAX-GalSim#29]
Hi, 

Using the drawing branch  I experience some problems that I expose here. Some can find a temporary workaround
 but it might be difficult for the rest of the soft development.

Let us try to make 2 galaxy profiles using a vmap on their flux & sigma parameters:

```python
def gen_gal(p:dict):
  gal = galsim.Gaussian(flux=p['flux'], sigma=p['sigma'])
  return gal

print("tst vmap gen_gal")
params = {'flux': jnp.array([1.e5, 2.e5]), 'sigma': jnp.array([1.,2.])}
res = jax.vmap(gen_gal)(params)
print(res)
```
the final error message  is 
```
--------------------


The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)

[/usr/local/lib/python3.8/dist-packages/jax_galsim/gaussian.py](https://localhost:8080/#) in __init__(self, half_light_radius, sigma, fwhm, flux, gsparams)
     63             super().__init__(sigma=sigma, flux=flux, gsparams=gsparams)
     64 
---> 65         self._sigsq = self.sigma**2
     66         self._inv_sigsq = 1.0 / self._sigsq
     67         self._norm = self.flux * self._inv_sigsq * Gaussian._inv_twopi

TypeError: unsupported operand type(s) for ** or pow(): 'object' and 'int'
```
See [Here a nb on Colab](https://colab.research.google.com/drive/1uy3RJnwpdB2WQ5GDBjKjnRoa6uiQh_H7?usp=sharing)

**Comment by jecampagne** [comment]
More 
# When I call to instantiate a single gaussian profile
JEC DBG: type self.sigma <class 'float'>
tst single: gen_gal galsim.Gaussian(sigma=1.0, flux=100000.0)

# Now on a vmap which aims to build 2 gal profiles
JEC DBG: type self.sigma <class 'jax.interpreters.batching.BatchTracer'>
JEC DBG: type self.sigma <class 'object'>

**Comment by jecampagne** [comment]
Some more input:
On the "drawing" branch
```python
def gen_gal(p:dict):
  gal = galsim.Gaussian(flux=p['flux'], sigma=p['sigma'])
  return gal.flux + gal.sigma

print("tst vmap gen_gal")
params = {'flux': jnp.array([1.e5, 2.e5]), 'sigma': jnp.array([1.,2.])}
res = jax.vmap(gen_gal)(params)
print(res)
```
gives as expected `[100001. 200002.]`.

But 
```python
def draw_gal_img(p:dict):
  #Define the galaxy profile.
  gal = galsim.Gaussian(flux=p['flux'], sigma=p['sigma'])
  img = gal.drawImage(scale=0.2)
  return img.array

params = {'flux': jnp.array([1.e5, 2.e5]), 'sigma': jnp.array([1.,2.])}
imgs = jax.vmap(draw_gal_img)(params)
print("result: ", imgs)
```
gives pbs in the `image.py` L156 due to
```
   if ncol != int(ncol) or nrow != int(nrow): 
```

Then, if one comment this line as well as those using `int`function  then on gets this Traceback

```
Traceback (most recent call last):
  File "test_vmap.py", line 33, in <module>
    imgs = jax.vmap(draw_gal_img)(params)
  File "/sps/lsst/users/campagne/anaconda3/envs/jaxgalsim/lib/python3.8/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/sps/lsst/users/campagne/anaconda3/envs/jaxgalsim/lib/python3.8/site-packages/jax/_src/api.py", line 1682, in vmap_f
    out_flat = batching.batch(
  File "/sps/lsst/users/campagne/anaconda3/envs/jaxgalsim/lib/python3.8/site-packages/jax/linear_util.py", line 167, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "test_vmap.py", line 27, in draw_gal_img
    img = gal.drawImage(scale=0.2)
  File "/sps/lsst/users/campagne/MyGalSim/JAX-GalSim/jax_galsim/gsobject.py", line 479, in drawImage
    image = prof._setup_image(image, nx, ny, bounds, add_to_image, dtype, center)
  File "/sps/lsst/users/campagne/MyGalSim/JAX-GalSim/jax_galsim/gsobject.py", line 282, in _setup_image
    image = Image(N, N, dtype=dtype)
  File "/sps/lsst/users/campagne/MyGalSim/JAX-GalSim/jax_galsim/image.py", line 160, in __init__
    self._array = self._make_empty(shape=(nrow, ncol), dtype=self._dtype)
  File "/sps/lsst/users/campagne/MyGalSim/JAX-GalSim/jax_galsim/image.py", line 440, in _make_empty
    return jnp.zeros(shape=shape, dtype=dtype)
  File "/sps/lsst/users/campagne/anaconda3/envs/jaxgalsim/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 2098, in zeros
    shape = canonicalize_shape(shape)
  File "/sps/lsst/users/campagne/anaconda3/envs/jaxgalsim/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 99, in canonicalize_shape
    return core.canonicalize_shape(shape, context)  # type: ignore
  File "/sps/lsst/users/campagne/anaconda3/envs/jaxgalsim/lib/python3.8/site-packages/jax/core.py", line 1899, in canonicalize_shape
    raise _invalid_shape_error(shape, context)
jax._src.traceback_util.UnfilteredStackTrace: TypeError: Shapes must be 1D sequences of concrete values of integer type, got (Traced<ShapedArray(int64[])>with<BatchTrace(level=1/0)> with
  val = DeviceArray([ 62, 120], dtype=int64)
  batch_dim = 0, Traced<ShapedArray(int64[])>with<BatchTrace(level=1/0)> with
  val = DeviceArray([ 62, 120], dtype=int64)
  batch_dim = 0).
If using `jit`, try using `static_argnums` or applying `jit` to smaller subfunctions.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

```

**Comment by jecampagne** [comment]
Now, fixing the size of the image can be a work around at the level of the drawing branch where there is no Convolution especially in the Fourier Space.
But, dealing with drawFFT, the first method in GalSim is the drawFFT_makeKImage. 
At the end there is a ImageCF (in single-precision) creation
To mimic 
```python
def f(n):
  dk = 2.*jnp.pi / (n * 0.2)
  bounds = galsim.BoundsI(0,n//2,-n//2,n//2)
  kimage = galsim.ImageCF(bounds=bounds, scale=dk)  
  return kimage
  ```
is ok for a single call
```python
f(128)
```
leads 
```
galsim.Image(bounds=galsim.BoundsI(xmin=DeviceArray(0, dtype=int64), xmax=DeviceArray(64, dtype=int64), ymin=DeviceArray(-64, dtype=int64), ymax=DeviceArray(64, dtype=int64)), array=
array([[0.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
       ...,
       [0.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j],
       [0.+0.j, 0.+0.j, 0.+0.j, ..., 0.+0.j, 0.+0.j, 0.+0.j]],
      dtype=complex64), wcs=galsim.PixelScale(DeviceArray(0.24543693, dtype=float64, weak_type=True)))
```  
But 
```python
jax.vmap(f)(jnp.array([128,128]))
```
leads to
```
---------------------------------------------------------------------------

UnfilteredStackTrace                      Traceback (most recent call last)

[<ipython-input-13-77dc35b25c9e>](https://localhost:8080/#) in <module>
----> 1 jax.vmap(f)(jnp.array([128,128]))

16 frames

[/usr/local/lib/python3.8/dist-packages/jax/_src/traceback_util.py](https://localhost:8080/#) in reraise_with_filtered_traceback(*args, **kwargs)
    161     try:
--> 162       return fun(*args, **kwargs)
    163     except Exception as e:

[/usr/local/lib/python3.8/dist-packages/jax/_src/api.py](https://localhost:8080/#) in vmap_f(*args, **kwargs)
   1681                   _mapped_axis_size(fun, in_tree, args_flat, in_axes_flat, "vmap"))
-> 1682     out_flat = batching.batch(
   1683         flat_fun, axis_name, axis_size_, in_axes_flat,

[/usr/local/lib/python3.8/dist-packages/jax/linear_util.py](https://localhost:8080/#) in call_wrapped(self, *args, **kwargs)
    166     try:
--> 167       ans = self.f(*args, **dict(self.params, **kwargs))
    168     except:

[<ipython-input-10-b8956b9d9f01>](https://localhost:8080/#) in f(n)
      3   bounds = galsim.BoundsI(0,n//2,-n//2,n//2)
----> 4   kimage = galsim.ImageCF(bounds=bounds, scale=dk)
      5   return kimage

[/usr/local/lib/python3.8/dist-packages/jax_galsim/image.py](https://localhost:8080/#) in ImageCF(*args, **kwargs)
    990     kwargs["dtype"] = jnp.complex64
--> 991     return Image(*args, **kwargs)
    992 

[/usr/local/lib/python3.8/dist-packages/jax_galsim/image.py](https://localhost:8080/#) in __init__(self, *args, **kwargs)
    162                 raise TypeError("bounds must be a galsim.BoundsI instance")
--> 163             self._array = self._make_empty(bounds.numpyShape(), dtype=self._dtype)
    164             self._bounds = bounds

[/usr/local/lib/python3.8/dist-packages/jax_galsim/image.py](https://localhost:8080/#) in _make_empty(self, shape, dtype)
    435         """Helper function to make an empty numpy array of the given shape."""
--> 436         return jnp.zeros(shape=shape, dtype=dtype)
    437 

[/usr/local/lib/python3.8/dist-packages/jax/_src/numpy/lax_numpy.py](https://localhost:8080/#) in zeros(shape, dtype)
   2097   lax_internal._check_user_dtype_supported(dtype, "zeros")
-> 2098   shape = canonicalize_shape(shape)
   2099   return lax.full(shape, 0, _jnp_dtype(dtype))

[/usr/local/lib/python3.8/dist-packages/jax/_src/numpy/lax_numpy.py](https://localhost:8080/#) in canonicalize_shape(shape, context)
     98   else:
---> 99     return core.canonicalize_shape(shape, context)  # type: ignore
    100 

[/usr/local/lib/python3.8/dist-packages/jax/core.py](https://localhost:8080/#) in canonicalize_shape(shape, context)
   1898     pass
-> 1899   raise _invalid_shape_error(shape, context)
   1900 

UnfilteredStackTrace: TypeError: Shapes must be 1D sequences of concrete values of integer type, got (Traced<ShapedArray(int64[])>with<BatchTrace(level=1/0)> with
  val = DeviceArray([129, 129], dtype=int64)
  batch_dim = 0, Traced<ShapedArray(int64[])>with<BatchTrace(level=1/0)> with
  val = DeviceArray([65, 65], dtype=int64)
  batch_dim = 0).
If using `jit`, try using `static_argnums` or applying `jit` to smaller subfunctions.

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------


The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)

[<ipython-input-13-77dc35b25c9e>](https://localhost:8080/#) in <module>
----> 1 jax.vmap(f)(jnp.array([128,128]))

[<ipython-input-10-b8956b9d9f01>](https://localhost:8080/#) in f(n)
      2   dk = 2.*jnp.pi / (n * 0.2)
      3   bounds = galsim.BoundsI(0,n//2,-n//2,n//2)
----> 4   kimage = galsim.ImageCF(bounds=bounds, scale=dk)
      5   return kimage

[/usr/local/lib/python3.8/dist-packages/jax_galsim/image.py](https://localhost:8080/#) in ImageCF(*args, **kwargs)
    989     """Alias for galsim.Image(..., dtype=numpy.complex64)"""
    990     kwargs["dtype"] = jnp.complex64
--> 991     return Image(*args, **kwargs)
    992 
    993 

[/usr/local/lib/python3.8/dist-packages/jax_galsim/image.py](https://localhost:8080/#) in __init__(self, *args, **kwargs)
    161             if not isinstance(bounds, BoundsI):
    162                 raise TypeError("bounds must be a galsim.BoundsI instance")
--> 163             self._array = self._make_empty(bounds.numpyShape(), dtype=self._dtype)
    164             self._bounds = bounds
    165             if init_value:

[/usr/local/lib/python3.8/dist-packages/jax_galsim/image.py](https://localhost:8080/#) in _make_empty(self, shape, dtype)
    434     def _make_empty(self, shape, dtype):
    435         """Helper function to make an empty numpy array of the given shape."""
--> 436         return jnp.zeros(shape=shape, dtype=dtype)
    437 
    438     def resize(self, bounds, wcs=None):

[/usr/local/lib/python3.8/dist-packages/jax/_src/numpy/lax_numpy.py](https://localhost:8080/#) in zeros(shape, dtype)
   2096     raise TypeError("expected sequence object with len >= 0 or a single integer")
   2097   lax_internal._check_user_dtype_supported(dtype, "zeros")
-> 2098   shape = canonicalize_shape(shape)
   2099   return lax.full(shape, 0, _jnp_dtype(dtype))
   2100 

[/usr/local/lib/python3.8/dist-packages/jax/_src/numpy/lax_numpy.py](https://localhost:8080/#) in canonicalize_shape(shape, context)
     97     return core.canonicalize_shape((shape,), context)  # type: ignore
     98   else:
---> 99     return core.canonicalize_shape(shape, context)  # type: ignore
    100 
    101 # Common docstring additions:

TypeError: Shapes must be 1D sequences of concrete values of integer type, got (Traced<ShapedArray(int64[])>with<BatchTrace(level=1/0)> with
  val = DeviceArray([129, 129], dtype=int64)
  batch_dim = 0, Traced<ShapedArray(int64[])>with<BatchTrace(level=1/0)> with
  val = DeviceArray([65, 65], dtype=int64)
  batch_dim = 0).
If using `jit`, try using `static_argnums` or applying `jit` to smaller subfunctions.
```

**Comment by ismael-mendoza** [comment]
Closed by #21
